GDJS/Runtime/howler-sound-manager/howler-sound-manager.ts

```diff
@@ -1,104 +1,296 @@
+     */
+    private _id?: integer;
 
-    constructor(o) {
-      super(o);
-      this._rate = o.rate || 1;
+    /**
+     * The Howl passed to the constructor.
+     * It defines the sound file that is being played.
+     */
+    private _howl: Howl;
 
-      //Add custom events listener to keep
-      //track of the sound status.
-      const that = this;
-      this.on('end', function () {
-        if (!that.loop()) {
-          that._canBeDestroyed = true;
-          that._paused = false;
-          that._stopped = true;
-        }
-      });
-      this.on('playerror', function (id, error) {
-        console.error(
-          "Can't play a sound, considering it as stopped. Error is:",
-          error
-        );
-        that._paused = false;
-        that._stopped = true;
-      });
+    /**
+     * An array of callbacks to call once the sound starts to play.
+     */
+    private _oncePlay: Array<HowlCallback> = [];
 
-      // Track play/pause event to be sure the status is
-      // sync'ed with the sound - though this should be redundant
-      // with `play`/`pause` methods already doing that. Keeping
-      // that to be sure that the status is always correct.
-      this.on('play', function () {
-        that._paused = false;
-        that._stopped = false;
-      });
-      this.on('pause', function () {
-        that._paused = true;
-        that._stopped = false;
-      });
+    /**
+     * An array of callbacks to call everytime the sound starts to play.
+     */
+    private _onPlay: Array<HowlCallback> = [];
+
+    constructor(howl: Howl) {
+      this._howl = howl;
     }
 
-    // Redefine `stop`/`play`/`pause` to ensure the status of the sound
-    // is immediately updated (so that calling `stopped` just after
-    // `play` will return false).
-    stop() {
-      this._paused = false;
-      this._stopped = true;
-      return super.stop();
+    /**
+     * Returns true if the associated howl is fully loaded.
+     */
+    isLoaded(): boolean {
+      return this._howl.state() === 'loaded';
     }
 
-    play() {
-      this._paused = false;
-      this._stopped = false;
-      return super.play();
+    /**
+     * Begins playback of the sound, or if the Howl is still loading, schedule playing for once it loads.
+     * @returns The current instance for chaining.
+     */
+    play(): this {
+      if (this.isLoaded()) {
+        const newID = this._howl.play(this._id === null ? undefined : this._id);
+        this._id = newID;
+
+        // Manually handle the play event before we have an ID.
+        // Before loading, howler won't register events as without an ID we cannot set a listener.
+        // Once we have an ID, we can transfer control of the events to howler.
+        // We also need to call them once as Howler doesn't for the first play event.
+        this._onPlay.forEach((func) => {
+          // Transfer the event to howler now that we have an ID
+          this.on('play', func);
```

Comment:
I'm not sure to understand this `this.on('play', func);`, can you explain again?
Why having a `func(newID)` below?

Reply:
Basically we want howler to handle the play events, triggered when the music starts playing or restarts because of the loop. Just doing this._howl.on("play") is fine or the latter, but not for the former, as the events trigger before we register the event (we may only register the event after getting the ID, which require playing which triggers the play event). To workaround that, before we have an ID, events are being held manually and we wait for the play to actually 
1. Call the functions to achieve the normal behavior of the play event being called when first starting the sound
2. Now that we have an ID actually set the callback so that the behavior of calling it again when it replay because of the loop

Technically that second point isn't necessary for how we internally use the play event, but I think it is better to keep the API consistent with the howler API to not confuse javascript users.